File: api/src/main/java/com/castlefrog/agl/domains/biniax/BiniaxSimulator.java
package com.castlefrog.agl.domains.biniax;

import java.util.ArrayList;
import java.util.List;

import com.castlefrog.agl.AbstractSimulator;
import com.castlefrog.agl.IllegalActionException;
import com.castlefrog.agl.TurnType;

/**
 * Biniax is a single agent stochastic domain.
 */
public final class BiniaxSimulator extends AbstractSimulator<BiniaxState, BiniaxAction> {
    private static final int N_AGENTS = 1;
    private static final TurnType TURN_TYPE = TurnType.SEQUENTIAL;
    private static final int INITIAL_ELEMENTS = 4;
    private static final int ELEMENT_INCREMENT_INTERVAL = 32;
    private static final double IMPASSIBLE_CHANCE = 0;

    private BiniaxSimulator(BiniaxState state) {
        legalActions_ = new ArrayList<>();
        legalActions_.add(new ArrayList<BiniaxAction>());
        setState(state);
        rewards_ = new int[] {1};
    }

    private BiniaxSimulator(BiniaxSimulator simulator) {
        super(simulator);
    }

    public BiniaxSimulator copy() {
        return new BiniaxSimulator(this);
    }

    public static BiniaxSimulator create(BiniaxState state) {
        return new BiniaxSimulator(state);
    }

    public void setState(BiniaxState state) {
        state_ = state;
        computeLegalActions();
    }

    public void stateTransition(List<BiniaxAction> actions) {
        BiniaxAction action = actions.get(0);
        if (!legalActions_.get(0).contains(action)) {
            throw new IllegalActionException(action, state_);
        }

        byte[][] locations = state_.getLocations();
        byte freeMoves = state_.getFreeMoves();
        int[] elementLocation = getElementLocation();
        int x = elementLocation[0];
        int y = elementLocation[1];
        byte element = state_.getLocation(x, y);
        int nTurns = state_.getNTurns();

        locations[x][y] = 0;
        switch (action) {
        case NORTH:
            y--;
            break;
        case EAST:
            x++;
            break;
        case SOUTH:
            y++;
            break;
        case WEST:
            x--;
            break;
        default:
            break;
        }

        if (locations[x][y] / BiniaxState.ELEMENT_LIMIT == element) {
            element = (byte) (locations[x][y] % BiniaxState.ELEMENT_LIMIT);
        } else if (locations[x][y] % BiniaxState.ELEMENT_LIMIT == element) {
            element = (byte) (locations[x][y] / BiniaxState.ELEMENT_LIMIT);
        }
        locations[x][y] = element;

        freeMoves--;
        if (freeMoves == 0) {
            freeMoves = BiniaxState.MAX_FREE_MOVES;
            // Move all elements down
            int emptyLocation = (int) (Math.random() * BiniaxState.getWidth());
            for (int i = BiniaxState.getHeight() - 1; i >= 0; i--) {
                for (int j = 0; j < BiniaxState.getWidth(); j++) {
                    if (i == 0) {
                        if (j != emptyLocation) {
                            if (Math.random() < IMPASSIBLE_CHANCE) {
                                locations[j][i] = -1;
                            } else {
                                locations[j][i] = (byte) generateRandomElementPair(nTurns);
                            }
                        } else {
                            locations[j][i] = 0;
                        }
                    } else {
                        locations[j][i] = locations[j][i - 1];
                    }
                }
            }
            // Move element back up if possible
            if (locations[x][y] == 0) {
                locations[x][y] = element;
                if (y < BiniaxState.getHeight() - 1) {
                    locations[x][y + 1] = 0;
                }
            } else if (locations[x][y] / BiniaxState.ELEMENT_LIMIT == element) {
                locations[x][y] = (byte) (locations[x][y] % BiniaxState.ELEMENT_LIMIT);
                if (y < BiniaxState.getHeight() - 1) {
                    locations[x][y + 1] = 0;
                }
            } else if (locations[x][y] % BiniaxState.ELEMENT_LIMIT == element) {
                locations[x][y] = (byte) (locations[x][y] / BiniaxState.ELEMENT_LIMIT);
                if (y < BiniaxState.getHeight() - 1) {
                    locations[x][y + 1] = 0;
                }
            }
        }
        state_ = new BiniaxState(locations, freeMoves, nTurns + 1);
        computeLegalActions();
    }

    /**
     * Creates a random element pair of dissimilar elements The elements are
     * always in order from smallest to largest
     * @return int of random values from 0 to numElements_ - 1
     */
    public static int generateRandomElementPair(int nTurns) {
        int nElementTypes = getNElementTypes(nTurns);
        int element1 = ((int) (Math.random() * nElementTypes)) + 1;
        int element2 = ((int) (Math.random() * (nElementTypes - 1))) + 1;
        if (element1 <= element2) {
            return element1 * BiniaxState.ELEMENT_LIMIT + element2 + 1;
        } else {
            return element2 * BiniaxState.ELEMENT_LIMIT + element1;
        }
    }

    private static int getNElementTypes(int nTurns) {
        return Math.min(INITIAL_ELEMENTS + nTurns / ELEMENT_INCREMENT_INTERVAL, BiniaxState.getMaxElements());
    }

    private int[] getElementLocation() {
        for (int i = 0; i < BiniaxState.getWidth(); i++) {
            for (int j = 0; j < BiniaxState.getHeight(); j++) {
                if (state_.getLocation(i, j) > 0 && state_.getLocation(i, j) < BiniaxState.ELEMENT_LIMIT) {
                    return new int[] {i, j};
                }
            }
        }
        throw new IllegalStateException("Element does not exist");
    }

    /**
     * A legal action is one that moves the single element to an empty space or
     * an element pair that contains that element and avoids being pushed off
     * the board.
     */
    private void computeLegalActions() {
        legalActions_.get(0).clear();
        int[] elementLocation = getElementLocation();
        int x = elementLocation[0];
        int y = elementLocation[1];
        int element = state_.getLocation(x, y);
        byte[][] locations = state_.getLocations();

        if (y != 0 && (locations[x][y - 1] == 0 ||
                        locations[x][y - 1] / BiniaxState.ELEMENT_LIMIT == element ||
                        locations[x][y - 1] % BiniaxState.ELEMENT_LIMIT == element)) {
            legalActions_.get(0).add(BiniaxAction.NORTH);
        }

        if (x != BiniaxState.getWidth() - 1) {
            int nextElement = 0;
            if (locations[x + 1][y] == 0) {
                nextElement = element;
            } else if (locations[x + 1][y] / BiniaxState.ELEMENT_LIMIT == element) {
                nextElement = locations[x + 1][y] % BiniaxState.ELEMENT_LIMIT;
            } else if (locations[x + 1][y] % 10 == element) {
                nextElement = locations[x + 1][y] / BiniaxState.ELEMENT_LIMIT;
            }

            if (nextElement != 0) {
                if (state_.getFreeMoves() > 1 ||
                        y < BiniaxState.getHeight() - 1 ||
                        locations[x + 1][y - 1] == 0 ||
                        locations[x + 1][y - 1] / BiniaxState.ELEMENT_LIMIT == nextElement ||
                        locations[x + 1][y - 1] % BiniaxState.ELEMENT_LIMIT == nextElement) {
                    legalActions_.get(0).add(BiniaxAction.EAST);
                }
            }
        }

        if (y != BiniaxState.getHeight() - 1 && (locations[x][y + 1] == 0 ||
                        locations[x][y + 1] / BiniaxState.ELEMENT_LIMIT == element ||
                        locations[x][y + 1] % BiniaxState.ELEMENT_LIMIT == element)) {
            legalActions_.get(0).add(BiniaxAction.SOUTH);
        }

        if (x != 0) {
            int nextElement = 0;
            if (locations[x - 1][y] == 0) {
                nextElement = element;
            } else if (locations[x - 1][y] / BiniaxState.ELEMENT_LIMIT == element) {
                nextElement = locations[x - 1][y] % BiniaxState.ELEMENT_LIMIT;
            } else if (locations[x - 1][y] % BiniaxState.ELEMENT_LIMIT == element) {
                nextElement = locations[x - 1][y] / BiniaxState.ELEMENT_LIMIT;
            }

            if (nextElement != 0) {
                if (state_.getFreeMoves() > 1 ||
                        y < BiniaxState.getHeight() - 1 ||
                        locations[x - 1][y - 1] == 0 ||
                        locations[x - 1][y - 1] / BiniaxState.ELEMENT_LIMIT == nextElement ||
                        locations[x - 1][y - 1] % BiniaxState.ELEMENT_LIMIT == nextElement) {
                    legalActions_.get(0).add(BiniaxAction.WEST);
                }
            }
        }
    }

    public int getNAgents() {
        return N_AGENTS;
    }

    public TurnType getTurnType() {
        return TURN_TYPE;
    }
}

File: api/src/main/java/com/castlefrog/agl/domains/biniax/BiniaxSimulator.kt
package com.castlefrog.agl.domains.biniax

import com.castlefrog.agl.Simulator
import java.util.ArrayList

/**
 * Biniax is a single agent stochastic domain
 */
class BiniaxSimulator(state: BiniaxState,
                      legalActions: List<MutableList<BiniaxAction>>? = null,
                      private val initialElements: Int,
                      private val maxElements: Int,
                      private val elementIncrementInterval: Int,
                      private val maxFreeMoves: Int) :
        Simulator<BiniaxState, BiniaxAction> {

    override var state: BiniaxState = state
        set(value) {
            field = value
            _legalActions = null
        }

    private var _legalActions: List<MutableList<BiniaxAction>>? = null
    override val legalActions: List<MutableList<BiniaxAction>>
        get() {
            if (_legalActions == null) {
                _legalActions = computeLegalActions(state)
            }
            return _legalActions ?: computeLegalActions(state)
        }

    override val rewards: IntArray = intArrayOf(1)

    init {
        _legalActions = legalActions
    }

    override fun copy(): BiniaxSimulator {
        return BiniaxSimulator(state.copy(), _legalActions?.copy(), initialElements, maxElements, elementIncrementInterval, maxFreeMoves)
    }

    override fun stateTransition(actions: List<BiniaxAction?>) {
        assert(actions.size == nAgents)
        val action = actions[0]
        if (action == null || !legalActions[0].contains(action)) {
            throw IllegalArgumentException("Illegal action, $action, from state, $state")
        }

        val locations = state.locations
        var freeMoves = state.freeMoves
        val elementLocation = state.elementLocation()
        var x = elementLocation[0]
        var y = elementLocation[1]
        var element = state.locations[x][y]
        val nTurns = state.nTurns

        locations[x][y] = 0
        when (action) {
            BiniaxAction.NORTH -> y--
            BiniaxAction.EAST -> x++
            BiniaxAction.SOUTH -> y++
            BiniaxAction.WEST -> x--
        }

        if (locations[x][y] / maxElements == element.toInt()) {
            element = (locations[x][y] % maxElements).toByte()
        } else if (locations[x][y] % maxElements == element.toInt()) {
            element = (locations[x][y] / maxElements).toByte()
        }
        locations[x][y] = element

        freeMoves--
        if (freeMoves.toInt() == 0) {
            freeMoves = maxFreeMoves.toByte()
            // Move all elements down
            val emptyLocation = (Math.random() * state.width).toInt()
            for (i in state.height - 1 downTo 0) {
                for (j in 0..state.width - 1) {
                    if (i == 0) {
                        if (j != emptyLocation) {
                            if (Math.random() < IMPASSIBLE_CHANCE) {
                                locations[j][i] = -1
                            } else {
                                locations[j][i] = generateRandomElementPair(initialElements, nTurns, maxElements,
                                        elementIncrementInterval).toByte()
                            }
                        } else {
                            locations[j][i] = 0
                        }
                    } else {
                        locations[j][i] = locations[j][i - 1]
                    }
                }
            }
            // Move element back up if possible
            if (locations[x][y].toInt() == 0) {
                locations[x][y] = element
                if (y < state.height - 1) {
                    locations[x][y + 1] = 0
                }
            } else if (locations[x][y] / maxElements == element.toInt()) {
                locations[x][y] = (locations[x][y] % maxElements).toByte()
                if (y < state.height - 1) {
                    locations[x][y + 1] = 0
                }
            } else if (locations[x][y] % maxElements == element.toInt()) {
                locations[x][y] = (locations[x][y] / maxElements).toByte()
                if (y < state.height - 1) {
                    locations[x][y + 1] = 0
                }
            }
        }
        state = BiniaxState(locations, maxElements, freeMoves, nTurns + 1)
    }

    override val nAgents: Int = 1

    companion object {
        private val IMPASSIBLE_CHANCE = 0.0
        private val DEFAULT_WIDTH = 5
        private val DEFAULT_HEIGHT = 7
        private val DEFAULT_INITIAL_ELEMENTS = 4
        private val DEFAULT_ELEMENT_INCREMENT_INTERVAL = 32
        private val DEFAULT_MAX_ELEMENTS = 10
        private val DEFAULT_BUFFER = 3
        private val DEFAULT_MAX_FREE_MOVES = 2

        fun create(width: Int = DEFAULT_WIDTH,
                   height: Int = DEFAULT_HEIGHT,
                   initialElements: Int = DEFAULT_INITIAL_ELEMENTS,
                   maxElements: Int = DEFAULT_MAX_ELEMENTS,
                   elementIncrementInterval: Int = DEFAULT_ELEMENT_INCREMENT_INTERVAL,
                   maxFreeMoves: Int = DEFAULT_MAX_FREE_MOVES,
                   buffer: Int = DEFAULT_BUFFER): BiniaxSimulator {
            return BiniaxSimulator(state = getInitialState(width, height, initialElements, maxElements,
                    elementIncrementInterval, maxFreeMoves, buffer),
                    initialElements = initialElements,
                    maxElements = maxElements,
                    elementIncrementInterval = elementIncrementInterval,
                    maxFreeMoves = maxFreeMoves)
        }

        fun getInitialState(width: Int = DEFAULT_WIDTH,
                            height: Int = DEFAULT_HEIGHT,
                            initialElements: Int = DEFAULT_INITIAL_ELEMENTS,
                            maxElements: Int = DEFAULT_MAX_ELEMENTS,
                            elementIncrementInterval: Int = DEFAULT_ELEMENT_INCREMENT_INTERVAL,
                            maxFreeMoves: Int = DEFAULT_MAX_FREE_MOVES,
                            buffer: Int = DEFAULT_BUFFER): BiniaxState {
            val locations = Array(width) { ByteArray(height) }
            for (i in 0..height - 1) {
                val emptyLocation = (Math.random() * width).toInt()
                for (j in 0..width - 1) {
                    if (j != emptyLocation && i < height - buffer) {
                        locations[j][i] = BiniaxSimulator.generateRandomElementPair(initialElements, 0, maxElements, elementIncrementInterval).toByte()
                        if (i == height - buffer - 1) {
                            locations[j][i] = (locations[j][i] % maxElements + maxElements).toByte()
                        }
                    }
                }
            }
            locations[width / 2][height - 1] = 1
            return BiniaxState(locations = locations, maxElements = maxElements, freeMoves = maxFreeMoves.toByte())
        }

        private fun BiniaxState.elementLocation(): IntArray {
            for (i in 0..width - 1) {
                for (j in 0..height - 1) {
                    if (locations[i][j] > 0 && locations[i][j] < maxElements) {
                        return intArrayOf(i, j)
                    }
                }
            }
            throw IllegalStateException("Element does not exist")
        }

        /**
         * A legal action is one that moves the single element to an empty space or
         * an element pair that contains that element and avoids being pushed off
         * the board.
         */
        private fun computeLegalActions(state: BiniaxState): List<MutableList<BiniaxAction>> {
            val legalActions = ArrayList<MutableList<BiniaxAction>>()
            legalActions.add(ArrayList<BiniaxAction>())
            val elementLocation = state.elementLocation()
            val x = elementLocation[0]
            val y = elementLocation[1]
            val element = state.locations[x][y].toInt()
            val locations = state.locations

            if (y != 0 && (locations[x][y - 1].toInt() == 0 ||
                    locations[x][y - 1] / state.maxElements == element ||
                    locations[x][y - 1] % state.maxElements == element)) {
                legalActions[0].add(BiniaxAction.NORTH)
            }

            if (x != state.width - 1) {
                var nextElement = 0
                if (locations[x + 1][y].toInt() == 0) {
                    nextElement = element
                } else if (locations[x + 1][y] / state.maxElements == element) {
                    nextElement = locations[x + 1][y] % state.maxElements
                } else if (locations[x + 1][y] % 10 == element) {
                    nextElement = locations[x + 1][y] / state.maxElements
                }

                if (nextElement != 0) {
                    if (state.freeMoves > 1 ||
                            y < state.height - 1 ||
                            locations[x + 1][y - 1].toInt() == 0 ||
                            locations[x + 1][y - 1] / state.maxElements == nextElement ||
                            locations[x + 1][y - 1] % state.maxElements == nextElement) {
                        legalActions[0].add(BiniaxAction.EAST)
                    }
                }
            }

            if (y != state.height - 1 && (locations[x][y + 1].toInt() == 0 ||
                    locations[x][y + 1] / state.maxElements == element ||
                    locations[x][y + 1] % state.maxElements == element)) {
                legalActions[0].add(BiniaxAction.SOUTH)
            }

            if (x != 0) {
                var nextElement = 0
                if (locations[x - 1][y].toInt() == 0) {
                    nextElement = element
                } else if (locations[x - 1][y] / state.maxElements == element) {
                    nextElement = locations[x - 1][y] % state.maxElements
                } else if (locations[x - 1][y] % state.maxElements == element) {
                    nextElement = locations[x - 1][y] / state.maxElements
                }

                if (nextElement != 0) {
                    if (state.freeMoves > 1 ||
                            y < state.height - 1 ||
                            locations[x - 1][y - 1].toInt() == 0 ||
                            locations[x - 1][y - 1] / state.maxElements == nextElement ||
                            locations[x - 1][y - 1] % state.maxElements == nextElement) {
                        legalActions[0].add(BiniaxAction.WEST)
                    }
                }
            }
            return legalActions
        }

        /**
         * Creates a random element pair of dissimilar elements The elements are
         * always in order from smallest to largest
         * @return int of random values from 0 to numElements_ - 1
         */
        fun generateRandomElementPair(initialElements: Int,
                                      nTurns: Int,
                                      maxElements: Int,
                                      elementIncrementInterval: Int): Int {
            val nElementTypes = getNElementTypes(initialElements, nTurns, maxElements, elementIncrementInterval)
            val element1 = (Math.random() * nElementTypes).toInt() + 1
            val element2 = (Math.random() * (nElementTypes - 1)).toInt() + 1
            if (element1 <= element2) {
                return element1 * maxElements + element2 + 1
            } else {
                return element2 * maxElements + element1
            }
        }

        private fun getNElementTypes(initialElements: Int,
                                     nTurns: Int,
                                     maxElements: Int,
                                     elementIncrementInterval: Int): Int {
            return Math.min(initialElements + nTurns / elementIncrementInterval, maxElements - 1)
        }
    }
}
